### PR TITLE
[Inductor][CPP] generalize the wgt tensor delete

### PR DESCRIPTION
**Pull request by leslie-fang-intel**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #135101

**Summary**
Previously, we assumed the packed weight for (`MKL/MKLDNN`) linear operations was at `new_input_nodes[1]`. However, this is not the case for `MKL linear`, where `new_input_nodes[1]` contains the original weight instead of the packed weight. To generalize the code, in this PR, we identify nodes that are present in `input_nodes` but not in `new_input_nodes`—indicating they are no longer used by the GEMM template and can be considered candidates for deletion.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

**Comment by leslie-fang-intel** [comment]
TAS test (BF16 Dynamic shape/FP32 Static shape) shows no extra failure.

File: torch/_inductor/codegen/cpp_gemm_template.py [hunk 1]
```diff
@@ -660,71 +660,112 @@ def preprocessor(inputs, layout):
...
-                                    )
-                                )
-                            )
-                        ):
-                            W_tensor_users += 1
+                        if share_storage(candidate_tensor, comp_tensor):
+                            candidate_tensor_users += 1
 
                 for node in reversed(V.graph.graph.nodes):
                     # The wgt tensor has been used by only 1 get_attr node
```
Comment (jgong5) [inline torch/_inductor/codegen/cpp_gemm_template.py:737]:
```suggestion
                    # The candidate tensor has been used by only 1 get_attr node
```

File: torch/_inductor/codegen/cpp_gemm_template.py [hunk 2]
```diff
@@ -660,71 +660,112 @@ def preprocessor(inputs, layout):
...
+                ):
+                    for new_node in new_input_nodes:
+                        if (
+                            isinstance(new_node, (ir.TensorBox, ir.StorageBox))
+                            and new_node.get_name() in V.graph.constants
+                        ):
+                            candidate_tensor = V.graph.constants[
+                                candidate_node.get_name()
+                            ]
+                            comp_tensor = V.graph.constants[new_node.get_name()]
```
Comment (jgong5) [inline torch/_inductor/codegen/cpp_gemm_template.py:699]: To name things consistently?
```suggestion
                            new_tensor = V.graph.constants[new_node.get_name()]
```

File: torch/_inductor/codegen/cpp_gemm_template.py [hunk 3]
```diff
@@ -660,71 +660,112 @@ def preprocessor(inputs, layout):
...
-                new_input_nodes, _ = reorder_and_filter(input_nodes, layout)
-
-                W_node = new_input_nodes[1]
-                assert W_node.get_name() in V.graph.constants
-                W = V.graph.constants[W_node.get_name()]
-                new_input_nodes[1] = W
-                new_input_nodes, _ = pack_weight(
-                    *normalize_shapes(*maybe_to_dense(new_input_nodes, layout))
+        def prune_tensors(input_nodes, new_input_nodes):
+            def share_storage(base_tensor: torch.Tensor, comp_tensor: torch.Tensor):
```
Comment (jgong5) [inline torch/_inductor/codegen/cpp_gemm_template.py:664]: perhaps we can factor it out and shared by other code?

**Review by jgong5 — CHANGES_REQUESTED** [review]
minor comments on naming, others LGTM.

File: torch/_inductor/codegen/cpp_gemm_template.py (hunk 1 above)
Comment (leslie-fang-intel) [reply torch/_inductor/codegen/cpp_gemm_template.py:737]: Thanks and changed.

File: torch/_inductor/codegen/cpp_gemm_template.py (hunk 2 above)
Comment (leslie-fang-intel) [reply torch/_inductor/codegen/cpp_gemm_template.py:699]: Thanks and changed.

**Review by jgong5 — APPROVED** [review]

**Comment by leslie-fang-intel** [comment]
Hi @jansel, could you kindly help to take a look of this PR?

**Review by jansel — APPROVED** [review]

**Pull request closed without merging**